src/backend/cuda/nvrtc/cache.cpp

```diff
@@ -378,6 +418,80 @@ string toString(const char *str) {
     return string(str);
 }
 
+template<>
+string toString(af_interp_type p) {
+    const char *retVal = NULL;
+#define CASE_STMT(v) \
+    case v: retVal = #v; break
+    switch (p) {
+        CASE_STMT(AF_INTERP_NEAREST);
+        CASE_STMT(AF_INTERP_LINEAR);
+        CASE_STMT(AF_INTERP_BILINEAR);
+        CASE_STMT(AF_INTERP_CUBIC);
+        CASE_STMT(AF_INTERP_LOWER);
+#if AF_API_VERSION >= 34
+        CASE_STMT(AF_INTERP_LINEAR_COSINE);
+        CASE_STMT(AF_INTERP_BILINEAR_COSINE);
+        CASE_STMT(AF_INTERP_BICUBIC);
+        CASE_STMT(AF_INTERP_CUBIC_SPLINE);
+        CASE_STMT(AF_INTERP_BICUBIC_SPLINE);
+#endif
+    }
+#undef CASE_STMT
+    return retVal;
+}
+
+template<>
+string toString(af_border_type p) {
+    const char *retVal = NULL;
+#define CASE_STMT(v) \
+    case v: retVal = #v; break
+    switch (p) {
+        CASE_STMT(AF_PAD_ZERO);
+        CASE_STMT(AF_PAD_SYM);
+        CASE_STMT(AF_PAD_CLAMP_TO_EDGE);
+    }
+#undef CASE_STMT
+    return retVal;
+}
+
+#if AF_API_VERSION >= 34
+template<>
+string toString(af_moment_type p) {
```

Comment:
should be defined in the moments file don't you think?

Reply:
It may make sense if you are just considering moments enum, but border type enum is used in multiple locations - it would be confusing in such cases. Right now, these specializations are centrally located in cache.cpp close to the general implementation.

I might move these later anyway when OpenCL cache uses similar mechanism to cache compiled kernels.